### ISSUE TITLE
Method Illuminate\Database\Eloquent\Collection::getKey does not exist

### ISSUE THREAD

**Issue by tillkruss** [laravel/nova-issues#3469]
- Laravel Version: v8.49.2
- Nova Version: 3.27.0
- PHP Version: 8.0.7
- Database Driver & Version: PostgreSQL 13.3

## Description

I'm able to attach a `Program` to a `Collection` from:

```
Nova > Resources > Programs > Collections > Attach Collection
```

However, when I try to attach a `Program` to a `Collection` from:

```
Nova > Collections > Programs > Attach Program
```

the request fails with:

```
[2021-07-07 21:50:54] local.ERROR: Method Illuminate\Database\Eloquent\Collection::getKey does not exist. {"userId":1,"exception":"[object] (BadMethodCallException(code: 0): Method Illuminate\\Database\\Eloquent\\Collection::getKey does not exist. at /vendor/laravel/framework/src/Illuminate/Macroable/Traits/Macroable.php:103)
[stacktrace]
#0 /vendor/laravel/nova/src/Rules/Relatable.php(96): Illuminate\\Support\\Collection->__call('getKey', Array)
#1 /vendor/laravel/nova/src/Rules/Relatable.php(63): Laravel\\Nova\\Rules\\Relatable->relationshipIsFull(Object(App\\Models\\Program), 'programs', '1')
#2 /vendor/laravel/framework/src/Illuminate/Validation/Validator.php(761): Laravel\\Nova\\Rules\\Relatable->passes('programs', '1')
#3 /vendor/laravel/framework/src/Illuminate/Validation/Validator.php(565): Illuminate\\Validation\\Validator->validateUsingCustomRule('programs', '1', Object(Laravel\\Nova\\Rules\\RelatableAttachment))
#4 /vendor/laravel/framework/src/Illuminate/Validation/Validator.php(394): Illuminate\\Validation\\Validator->validateAttribute('programs', Object(Laravel\\Nova\\Rules\\RelatableAttachment))
#5 /vendor/laravel/framework/src/Illuminate/Validation/Validator.php(425): Illuminate\\Validation\\Validator->passes()
#6 /vendor/laravel/framework/src/Illuminate/Validation/Validator.php(468): Illuminate\\Validation\\Validator->fails()
#7 /vendor/laravel/nova/src/Http/Controllers/ResourceAttachController.php(62): Illuminate\\Validation\\Validator->validate()
#8 /vendor/laravel/framework/src/Illuminate/Support/helpers.php(263): Laravel\\Nova\\Http\\Controllers\\ResourceAttachController->Laravel\\Nova\\Http\\Controllers\\{closure}(Array)
#9 /vendor/laravel/nova/src/Http/Controllers/ResourceAttachController.php(65): tap(Array, Object(Closure))
#10 /vendor/laravel/nova/src/Http/Controllers/ResourceAttachController.php(32): Laravel\\Nova\\Http\\Controllers\\ResourceAttachController->validate(Object(Laravel\\Nova\\Http\\Requests\\NovaRequest), Object(App\\Models\\Collection), 'App\\\\Nova\\\\Collec...')
#11 /vendor/laravel/framework/src/Illuminate/Routing/Controller.php(54): Laravel\\Nova\\Http\\Controllers\\ResourceAttachController->handle(Object(Laravel\\Nova\\Http\\Requests\\NovaRequest), 'collections', '1', 'programs')"} 
```

## Detailed steps to reproduce the issue on a fresh Nova installation

These are the models:

```php
class Collection extends Model
{
    public function programs()
    {
        return $this->morphedByMany(Program::class, 'collectable');
    }
}

class Program extends Model
{
    public function collections()
    {
        return $this->morphToMany(Collection::class, 'collectable');
    }
}
```

These are the Nova resources:

```php
class Program extends Resource
{
    public function fields(Request $request)
    {
        return [
            MorphToMany::make('Collections'),
        ];
    }
}

class Collection extends NovaResource
{
    public function fields(Request $request)
    {
        return [
            MorphedByMany::make('Programs'),
        ];
    }
}
```

**Comment by crynobone** [comment]
I tried to replicate the issue but unable to get the same error: https://github.com/nova-issues/issue-3469/commit/546b2180e5e34bd128c418e579bf1caa8fcb637c

This also been tested with Postgres (using Takeout) with the same result:

```ini
DB_CONNECTION=pgsql
DB_HOST=127.0.0.1
DB_PORT=5432
DB_DATABASE=issue_3469
DB_USERNAME=postgres
DB_PASSWORD=password
```

https://user-images.githubusercontent.com/172966/124843500-a3d28900-dfc4-11eb-96f6-16591ff34580.mp4

**Comment by crynobone** [comment]
Do you by any change has another `MorphOne` field under `Collection` resource with attribute `programs`. This could cause the issue.

**Comment by tillkruss** [comment]
Yes, the full resource is:

```php
class Collection extends NovaResource
{
    public function fields(Request $request)
    {
        return [
            ID::make()->sortable(),

            Text::make('Name')
                ->sortable()
                ->rules('required', 'max:255'),

            MorphTo::make('Collector')
                ->nullable()
                ->types([
                    Program::class,
                ]),

            MorphedByMany::make('Programs'),

            MorphedByMany::make('Resources'),

            MorphedByMany::make('Challenges'),
        ];
    }
}
```

And the model is:

```php
class Collection extends Model
{
    public function collector()
    {
        return $this->morphTo();
    }

    public function programs()
    {
        return $this->morphedByMany(Program::class, 'collectable')
            ->as('collectable')
            ->using(Collectable::class);
    }

    public function resources()
    {
        return $this->morphedByMany(Resource::class, 'collectable')
            ->using(Collectable::class);
    }

    public function challenges()
    {
        return $this->morphedByMany(Challenge::class, 'collectable')
            ->using(Collectable::class);
    }
}
```

**Comment by crynobone** [comment]
But none of the above would resolve the attribute as `programs` other than `MorphedByMany::make('Programs')`.

**Comment by tillkruss** [comment]
@crynobone: Are you able to reproduce the issue? Is there anything else I can provide to help resolve this?

**Comment by crynobone** [comment]
@tillkruss I made a PR to override `relationshipIsFull()` on `Laravel\Nova\Rules\RelatableAttachment` since it only being used from `BelongsToMany` and `MorphedToMany` field.

[1414.patch](https://github.com/laravel/nova-issues/files/6794593/1414.patch.txt)

However, I was unable to create reproducing code for this issue. Not sure what's still missing: https://github.com/nova-issues/issue-3469/commit/bf2b9ea39bf8a2cd3e62d9056a6c31e1e203d713

**Comment by tillkruss** [comment]
@crynobone: I just tested that patch and that fixes the bug on my local installation. Would you like me to invite you to the app? It's private but nothing secret.

**Comment by crynobone** [comment]
Fixed in v3.28.0: https://nova.laravel.com/releases/3.28.0
